### ISSUE TITLE
Make generated APIs easier to work with

### ISSUE THREAD

**Issue by DanielMSchmidt** [kubernetes-client/javascript#511]
**TL;DR** 

Move from positional arguments to object arguments for the sake of readability and extendability

**Details** 

The API that is currently generated is quite hard read if one has to fill in a property that is not the first or second argument. 

The current function signature is like this (on the example of namespaces)
```ts
export class CoreV1Api {
    public async createNamespace (body: V1Namespace, pretty?: string, dryRun?: string, fieldManager?: string, options: {headers: {[name: string]: string}} = {headers: {}}) : Promise<{ response: http.IncomingMessage; body: V1Namespace;  }>{}
	public async listNamespace (pretty?: string, allowWatchBookmarks?: boolean, _continue?: string, 	fieldSelector?: string, labelSelector?: string, limit?: number, resourceVersion?: string, timeoutSeconds?: number, watch?: boolean, options: {headers: {[name: string]: string}} = {headers: {}}) : Promise<{ response: http.IncomingMessage; body: V1NamespaceList;  }> {}
}
```

This leads to calls like this `api.createNamespace(namespace, undefined, true, undefined, {headers: {TRACE_REQUEST: "1"}})`, which are hard to understand / reason about without looking into the client itself.

I would like to propose a different format: 

```ts
// These types would be exported by the package and imported in this file
type Options = {
  pretty?: string
  dryRun?: string
  options?: {headers: {[name: string]: string}}
}
type Result<T> =  Promise<{ response: http.IncomingMessage; body: T;  }>

type CreateOptions<T> = Options & {
  body: T
  fieldManager?: string,
}
type CreateFunction<T> = (opts: CreateOptions<T>) => CreateResult<T>

type ListOptions = Options & {
	allowWatchBookmarks?: boolean
	_continue?: string
	fieldSelector?: string
	labelSelector?: string
	limit?: number
	resourceVersion?: string
	timeoutSeconds?: number
	watch?: boolean
}
type ListFunction<T> = (opts: ListOptions) => Result<V1NamespaceList>


// The generated file begins here
export class CoreV1Api {
    public async createNamespace (opts: CreateOptions<T>): CreateResult<T>{}
	public async listNamespace (opts: ListOptions) : Result<V1NamespaceList> {}
	// ...
}
```

This would allow us to call the api in a readable way: `api.createNamespace({body: namespace, dryRun: true, options: {headers: {TRACE_REQUEST: "1"}})`. Someone without intimate knowledge of the client library can instantly understand this.

Another benefit of having an object as configuration is that it easier to extend / change without breaking clients and that a user can define their own abstractions. For example defining a function that adds a header is a very different experience after we implement this:

```ts
type ListFn<T> = (pretty?: string, allowWatchBookmarks?: boolean, _continue?: string, 	fieldSelector?: string, labelSelector?: string, limit?: number, resourceVersion?: string, timeoutSeconds?: number, watch?: boolean, options: {headers: {[name: string]: string}} = {headers: {}}) : Promise<{ response: http.IncomingMessage; body: V1NamespaceList;) => T
function addHeader<T>(fn: ListFn<T> | CreateFn<T> | ...): T {
  if (isListFn(fn)) {
    return (pretty?: string, allowWatchBookmarks?: boolean, _continue?: string, 	fieldSelector?: string, labelSelector?: string, limit?: number, resourceVersion?: string, timeoutSeconds?: number, watch?: boolean, options: {headers: {[name: string]: string}} = {headers: {}}): T => fn(pretty, allowWatchBookmarks, _continue, fieldSelector, labelSelector, limit, resourceVersion, timeoutSeconds, watch, {...options, headers: {...options.headers, myHeader: "is there"})
  }

  if (isCreateFn(fn)){ .... }
}

// Versus after this change

function addHeader<T extends Options, R>(fn: (opts: T) => R): (opts: T) => R {
	return (opts) => fn({...opts, options: {...opts.options, headers: {...opts.options.headers, myHeader: "is there"}})
}
```

**Comment by brendandburns** [comment]
All of this code comes from the openapi code generator (`typescript-node`). If we want to make changes, the changes need to be made there.

Though for what it's worth, we need to also change away from the `request` generator, since the `request` library is deprecated.

**Comment by DanielMSchmidt** [comment]
I can take a look and tackle those issues 👍

**Comment by jgehrcke** [comment]
@DanielMSchmidt small world :-) Hope you're good!

> we need to also change away from the request generator, since the request library is deprecated.

Yes! And whichever HTTP client we choose (I think I can recommend `got` by now) we need to find a way to expose HTTP client options in a generic and yet TypeScript-supported way. Related: ttps://github.com/kubernetes-client/javascript/issues/544

**Comment by Maistho** [comment]
/reopen
/remove-lifecycle rotten

This is still very much relevant.

**Comment by drubin** [comment]
@brendandburns I am not sure how to tackle this but It think moving away from request lib should be done as its been deprecated for ages.

Do you think we can assign that to you?

**Comment by theomessin** [comment]
Could we not just switch over to `typescript-fetch`? It'd be two birds with one stone since it uses an options object instead of positional arguments (https://github.com/OpenAPITools/openapi-generator/pull/569).
